### PR TITLE
Feat 265 *platform wide* realtime support

### PR DESCRIPTION
**Pull request by eldadfux**
## What does this PR do?

Adding realtime server for allowing live updates and interactions with the Appwrite platform.

The server will include multiple realtime protocols support including (in order of priority):
- [x] Websocket - **in progress**
- [ ] MQTT
- [ ] Socket.io
- [ ] SSE

This PR is still a work in progress. Stuff still yet to be implemented:
- [ ] Abuse control check on connection start (X connections per Y hours, X connection per user)
- [ ] Limit payload size
- [ ] Defining message structure: { status: "ok"|"error", event: EVENT_NAME, data: <any arbitrary data> }
- [ ] JWT Authentication (passed in path / or in message) **related to #511 **
- [ ] List of channels
- [x] Channels subscriptions (per resource / group of resources (collections / files))
- [ ] Logout on current session delete event
- [ ] Integrate messaging worker for event triggering
- [ ] Add e2e tests
- [ ] Update installation script compose file

## Test Plan

TODO: add new set of tests to both test connection establishment and stream functionality. Trigger mock event, assert payload is correct on each listener.

## Related PRs and Issues

#511 
#265 
#509

### CODE REVIEW

File: app/controllers/api/health.php [hunk 1]
```diff
@@ -30,6 +30,19 @@
         $response->json(['version' => APP_VERSION_STABLE]);
     }, ['response']);
 
+App::get('/v1/health/realtime')
```
Comment (eldadfux) [inline app/controllers/api/health.php:33]: This is only used for testing and will be deleted later.

**Comment by TorstenDittmann** [comment]
May I ask where the need for multiple real-time protocols comes from?

**Comment by eldadfux** [comment]
@TorstenDittmann not much of a need, but more of a possibility. Seems to be pretty easy to support multiple protocols using Swoole.

This also have a very good fit with Appwrite agenda of being tech and platform agonistic, and might allow new use-cases. There were a few mentions in the community about MQTT and Websocket, I guess that will allow easier integration with Appwrite for platform that are already comfortable with a specific protocol or client SDK.

**Comment by TorstenDittmann** [comment]
I see nothing explicitly wrong with this code for now, since it is a starting point anyway. 

The plan is that all possible functions that are normally requested via HTTP are also accessible via the realtime protocol, right?

Also, have you figured out a plan to manage connections and subscriptions yet? 

Amazing work! 👍

**Comment by eldadfux** [comment]
> The plan is that all possible functions that are normally requested via HTTP are also accessible via the realtime protocol, right?

@TorstenDittmann I'm still thinking about it, not sure how the message request will look like, maybe something similar to GraphQL or GraphQL itself? I was thinking of starting with a read-only mode where users can subscribe and receive events and use it in combination with the REST API. What do you think?

> Also, have you figured out a plan to manage connections and subscriptions yet?

This is what I'm thinking about. I'm gonna use a hashmap of all the users where the key is going to be the connected user ID and the value is going to be a hashmap of channel names and the user connection ID, like this: `user_id` => [channels => ['channel1' => true, 'channel2' => true], connection => x].

Every new event message will do a hash lookup, which should be pretty fast, find all relevant users, check if they have subscribed to the messages, and then send the message if relevant (if the user has subscribed for it).

The performance here should be O(N), where N is the number of "read" permissions in a resource [file/document/collection] - (@TorstenDittmann let me know if I'm missing something). Maybe we should add a limit to the max number of permissions?

On each connection establishment, we'll verify the user has privileges to access that channel. We'll remove the connection ID on every connection close event or on specific failure errors.

For channel logic, I had this in mind: (but I'd love some feedback)

| Channel | Description |
|---|---|
| account | All account related events (session create, name update...) |
| collections | Any update/delete/create events for collections where user has read access |
| collections.[ID] | Any update/delete/create events to a given collection where user has read access |
| collections.[ID].documents | Any update/delete/create events to any document in a given collection where user has read access |
| documents | Any update/delete/create events to documents where user has read access |
| documents.[ID] | Any update/delete/create events to a given document where user has read access |
| files | Any update/delete/create events to file where user has read access |
| files.[ID] | Any update/delete/create events to a given file where user has read access |

**Comment by joshdvir** [comment]
@eldadfux I think that adding realtime feature to Appwrite will be a great addition, just a few things to keep in mind:

1. MQTT is not really a realtime protocol, it's a lightweight messaging protocol I'm not sure what use case are you trying to add here with it.
2. SSE only allows 1 way communication, only the server is sending messages to the client which can be limiting, I prefer Websockets by far.
3. I like the approach you took when you wanted to add a new feature you added a new component to complement that feature, I think this should also apply here, Scaling Websockets is different than scaling HTTP services so I think the component should be a new one that can be scaled independently and not part of the main application.
Moreover in the past I used https://centrifugal.github.io/centrifugo/ as a websockets server with much success and adding this type of component will save a lot of development time since the main application would just need to "manage" it and define the protocol to use.
Centrifugo also support SockJS which is like Socket.io.

Let me know what you think

Josh

**Comment by eldadfux** [comment]
Thanks for the feedback @joshdvir!

About the points you raised:
1. MQTT can be used to communicate between Appwrite and a developer custom backend located on the same network or datacenter. Also one of our main agendas is to be tech and platform agnostic in order to allow max customisation and flexibility with Appwrite.

2. Actually that can be good enough for a lot of use cases where you just need a listener. Even with Websocket will probably start by supporting on way communication and on phase 2 we'll add 2-way communication, but that hasn't been decided yet.

3. Completely agree. The realtime server is going to be an independent & stateless Docker container which will be super easy to both scale and debug. We're using Swoole for the Websocket server mainly because we already use it for HTTP (with great performance) and it allow us to easily reuse a lot of our codebase.

**Comment by eldadfux** [comment]
@TorstenDittmann are we OK to close this PR in favor of #948 ?

**Comment by TorstenDittmann** [comment]
> @TorstenDittmann are we OK to close this PR in favor of #948 ?

Yes, we can close this one.

**Pull request closed without merging**